### PR TITLE
Never follow bitcoind backwards

### PR DESCRIPTION
**Pull request by rustyrussell**
This solves the underlying reason why #3243 was such a problem!

### CODE REVIEW

File: tests/test_misc.py [hunk 1]
```diff
@@ -1145,6 +1145,57 @@ def test_rescan(node_factory, bitcoind):
...
+@pytest.mark.xfail(strict=True)
+def test_bitcoind_goes_backwards(node_factory, bitcoind):
+    """Check that we refuse to acknowledge bitcoind giving a shorter chain without explicit rescan"""
+    l1 = node_factory.get_node(may_fail=True, allow_broken_log=True)
+
+    bitcoind.generate_block(10)
+    sync_blockheight(bitcoind, [l1])
+    l1.stop()
+
+    # Now shrink chain (invalidateblock leaves 'headers' field)
```
Comment (cdecker) [inline tests/test_misc.py]: Just tried it myself, and indeed the `headers` remains unchanged after `invalidateblock`, however if `invalidateblock` is followed by a restart `headers` will reset to the expected value. Maybe that's a more stable way of going backwards than blowing away the underlying directory?

Comment (rustyrussell) [reply tests/test_misc.py]: Ah, that would have been easier!

**Comment by rustyrussell** [comment]
Rebased to use the Decker technique for creating reorgs...

File: tests/test_misc.py [hunk 2]
```diff
@@ -1145,6 +1145,60 @@ def test_rescan(node_factory, bitcoind):
...
+
+    bitcoind.rpc.invalidateblock(bitcoind.rpc.getblockhash(105))
+    bitcoind.rpc.stop()
+    TailableProc.stop(bitcoind)
+    bitcoind.start()
+    bitcoind.generate_block(5)
+
+    # It will ignore bitcoind and keep asking for block 110.
+    time.sleep(5)
+    assert l1.rpc.getinfo()['blockheight'] == 110
```
Comment (cdecker) [inline tests/test_misc.py]: This could be made into a `wait_for`. Happy to fix it in a cleanup though.

**Comment by cdecker** [comment]
Seems we have a bit of collateral damage:

 - `test_db_dangling_peer_fix` uses a DB that was synced up to 205 while we migrate on a chain that has 101 blocks, need to generate 104 more blocks before starting the node, or specify a rescan.
 - `test_rescan` seems to believe it was once synced to a 500k chain, and is being started with a 101 block chain.

Other than that LGTM :+1:

File: tests/test_misc.py (hunk 2 above)
Comment (rustyrussell) [reply tests/test_misc.py]: No, it can't: we want to make sure it's *still* at 110.

**Comment by rustyrussell** [comment]
> Seems we have a bit of collateral damage:
> 
>     * `test_db_dangling_peer_fix` uses a DB that was synced up to 205 while we migrate on a chain that has 101 blocks, need to generate 104 more blocks before starting the node, or specify a rescan.
> 
>     * `test_rescan` seems to believe it was once synced to a 500k chain, and is being started with a 101 block chain.
> 
> 
> Other than that LGTM +1

Thanks, fixed.

I actually like the change in behavior (refusing to start) if you ask it to rescan from a not-yet-present block, so I've kept and tested that.

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/pull/3274/commits/cdafcf35f8ccf88e789759c775f6d870a404144c

Fixed all the test collateral damage.

**Pull request merged**